motion polls

MaximilianKrambach opened the pull request:
Creating, displaying, editing, deleting polls for motions

- print ballots does not work yet (do we have a pdf creation yet?)
- some layout oddities
- I moved setting/displaying 'special values' into a separate branch (see [motion-poll-specialvalues](https://github.com/MaximilianKrambach/OpenSlides/tree/motionpoll-specialvalues) )

MaximilianKrambach commented on client/src/app/site/motions/components/motion-poll/motion-poll.component.spec.ts:
```diff
@@ -0,0 +1,30 @@
+import { MetaTextBlockComponent } from '../meta-text-block/meta-text-block.component';
+
+describe('MotionPollComponent', () => {
+    // TODO testing fails if personalNotesModule (also having the MetaTextBlockComponent)
```
Here I could use some help in angular magic

MaximilianKrambach replied:
After discussion I will comment this test for now, and we should collect an issue with missing/failing tests due to async/angular magic

MaximilianKrambach commented:
sorry for the after-commits; I discovered some issues and undocumented bits while working on something else. State is still 'reviewable' (and help needed for that testing behaviour) @tsiegleauq

MaximilianKrambach commented:
So, now the (new) linter is happy again, and there is only that failing test mentioned above that's left

MaximilianKrambach commented:
Squished and rebased on master, again @emanuelschuetze @tsiegleauq

tsiegleauq commented on client/src/app/core/poll.service.ts:
```diff
@@ -0,0 +1,113 @@
+     * An array of value - label pairs for special value signifiers.
+     * TODO: Should be given by the server, and editable. For now: hard coded
+     */
+    private _specialPollVotes: [number, string][] = [[-1, 'Majority'], [-2, 'Undocumented']];
```
the config varbiable should already exist.
For motion poll it would be:
`motions_poll_default_majority_method`

MaximilianKrambach replied:
That is something else (the choice if the default method used for the quorum calculations). I need the 'special values that are not really numbers'.

tsiegleauq replied:
OK

tsiegleauq commented on client/src/app/core/poll.service.ts:
```diff
@@ -0,0 +1,113 @@
+
+    /**
+     * Gets an icon for a Poll Key
+     * @param key
```
missing `@returns` statement

tsiegleauq commented on client/src/app/core/poll.service.ts:
```diff
@@ -0,0 +1,113 @@
+    /**
+     * getter for the special votes
+     */
+    public get specialPollVotes(): [number, string][] {
```
missing `@returns` statement

tsiegleauq commented on client/src/app/core/poll.service.ts:
```diff
@@ -0,0 +1,113 @@
+    /**
+     * Gets a label for a poll Key
+     */
+    public getLabel(key: CalculablePollKey): string {
```
missing `@returns` statement

tsiegleauq commented on client/src/app/core/poll.service.ts:
```diff
@@ -0,0 +1,113 @@
+     * representation of themselves
+     * @param value
+     */
+    public getSpecialLabel(value: number): string {
```
missing `@returns` statement

tsiegleauq commented on client/src/app/core/services/constants.service.ts:
```diff
@@ -78,6 +78,7 @@ export class ConstantsService extends OpenSlidesComponent {
      */
     public get(key: string): Observable<any> {
         if (this.constants) {
+            console.log(this.constants);
```
please remove

tsiegleauq commented on client/src/app/site/motions/components/motion-poll/motion-poll-dialog.component.ts:
```diff
@@ -0,0 +1,75 @@
+     * Constructor. Retrieves necessary metadata from the pollService,
+     * injects the poll itself
+     */
+    public constructor(
```
doc missing

tsiegleauq commented on client/src/app/site/motions/components/motion-poll/motion-poll.component.ts:
```diff
@@ -0,0 +1,232 @@
+import { MotionPollDialogComponent } from './motion-poll-dialog.component';
+import { MotionRepositoryService } from '../../services/motion-repository.service';
+
+@Component({
```
doc

tsiegleauq commented on client/src/app/site/motions/components/motion-poll/motion-poll.component.ts:
```diff
@@ -0,0 +1,232 @@
+    styleUrls: ['./motion-poll.component.scss']
+})
+export class MotionPollComponent implements OnInit {
+    public pollValues: CalculablePollKey[];
```
doc

tsiegleauq commented on client/src/app/site/motions/services/motion-poll.service.ts:
```diff
@@ -0,0 +1,174 @@
+})
+export class MotionPollService extends PollService {
+    /**
+     * list of
```
lost of what? ;-)

tsiegleauq commented on client/src/app/site/motions/services/motion-poll.service.ts:
```diff
@@ -0,0 +1,174 @@
+     */
+    public pollValues: CalculablePollKey[] = ['yes', 'no', 'abstain', 'votesvalid', 'votesinvalid', 'votescast'];
+
+    public constructor(config: ConfigService) {
```
doc

tsiegleauq commented on client/src/app/site/motions/components/motion-poll/motion-poll.component.ts:
```diff
@@ -0,0 +1,232 @@
+                }
+            }
+        });
+        this.config.get('motions_poll_default_majority_method').subscribe(method => {
```
exists in the service as well. You might just want to access the service here

tsiegleauq commented on client/src/app/site/motions/services/motion-poll.service.ts:
```diff
@@ -0,0 +1,174 @@
+                }
+                return poll.yes + poll.no;
+            default:
+                throw new Error('TODO lazy programmer');
```
?

tsiegleauq commented on client/src/app/site/motions/services/motion-poll.service.ts:
```diff
@@ -0,0 +1,174 @@
+
+    /**
+     * Calculates which number is needed for the quorum to be surpassed
+     * @returns the first integer number larger than the required majority,
```
returns under @param documentation

tsiegleauq commented on client/src/app/site/motions/services/motion-repository.service.ts:
```diff
@@ -632,4 +633,34 @@ export class MotionRepositoryService extends BaseRepository<ViewMotion, Motion>
+        await this.httpService.post(url);
+    }
+
+    public async updatePoll(poll: MotionPoll): Promise<void> {
```
doc

tsiegleauq commented on client/src/app/site/motions/services/motion-repository.service.ts:
```diff
@@ -632,4 +633,34 @@ export class MotionRepositoryService extends BaseRepository<ViewMotion, Motion>
+        await this.httpService.put(url, data);
+    }
+
+    public async deletePoll(poll: MotionPoll): Promise<void> {
```
doc

MaximilianKrambach commented:
I addressed the documentation issues and implemented a bit of design feedback I got per mail. However, design is still temporary.

Also, I re-added the 'special values' message from 2.3, as the special value treatment with buttons proves to be a bit more complex (and will be done later).

@tsiegleauq  the last two commits are reviewable again

tsiegleauq commented on client/src/app/site/motions/components/motion-detail/motion-detail.component.html:
```diff
@@ -309,8 +309,8 @@ <h4 translate>Origin</h4>
             </os-motion-poll>
-            <mat-divider></mat-divider>
             <button mat-button *ngIf="perms.isAllowed('createpoll', motion)" (click)="createPoll()">
+                <mat-icon class="main-nav-color"> poll</mat-icon>
```
one whitespace too much

tsiegleauq commented on client/src/app/site/motions/components/motion-poll/motion-poll-dialog.component.html:
```diff
@@ -1,13 +1,20 @@
-    <button mat-button (click)="submit()" translate>Ok</button>
+<div class="submit-buttons">
     <button mat-button (click)="cancel()" translate>Cancel</button>
+    <button mat-button (click)="submit()" translate>Ok</button>
```
should be named "Save",
should be before cancel and should be aligned left

MaximilianKrambach commented:
@tsiegleauq rebased and your last remarks addressed
